# Naives Bayes Algorithm

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import BernoulliNB, GaussianNB, MultinomialNB
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

df = pd.read_csv('https://raw.githubusercontent.com/4GeeksAcademy/naive-bayes-project-tutorial/main/playstore_reviews.csv', delimiter=',')

df.head()

,package_name,review,polarity
0,com.facebook.katana,privacy at least put some option appear offli...,0
1,com.facebook.katana,"messenger issues ever since the last update, ...",0
2,com.facebook.katana,profile any time my wife or anybody has more ...,0
3,com.facebook.katana,the new features suck for those of us who don...,0
4,com.facebook.katana,forced reload on uploading pic on replying co...,0


In [ ]:
df.describe()

,polarity
count,891.000000
mean,0.344557
std,0.475490
min,0.000000
25%,0.000000
50%,0.000000
75%,1.000000
max,1.000000


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   package_name  891 non-null    object
 1   review        891 non-null    object
 2   polarity      891 non-null    int64 
dtypes: int64(1), object(2)
memory usage: 21.0+ KB


In [ ]:
df.drop(["package_name"], axis=1, inplace=True)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   review    891 non-null    object
 1   polarity  891 non-null    int64 
dtypes: int64(1), object(1)
memory usage: 14.1+ KB


In [ ]:
df["review"] = df["review"].str.strip().str.lower()

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(df["review"], df["polarity"], test_size=0.2, random_state=42)

X_train.to_csv("../data/processed/X_train_ts_02_rs_42.csv", index=False)
y_train.to_csv("../data/processed/y_train_ts_02_rs_42.csv", index=False)
X_test.to_csv("../data/processed/X_test_ts_02_rs_42.csv", index=False)
y_test.to_csv("../data/processed/y_test_ts_02_rs_42.csv", index=False)

X_train.head()

331    just did the latest update on viber and yet ag...
733    keeps crashing it only works well in extreme d...
382    the fail boat has arrived the 6.0 version is t...
704    superfast, just as i remember it ! opera mini ...
813    installed and immediately deleted this crap i ...
Name: review, dtype: object

In [7]:
vec_model = CountVectorizer(stop_words = 'english')
X_train = vec_model.fit_transform(X_train).toarray()
X_test = vec_model.transform(X_test).toarray()

In [8]:
b_model = BernoulliNB(alpha=0.1)
b_model.fit(X_train, y_train)

BernoulliNB(alpha=0.1)

In [9]:
y_pred = b_model.predict(X_test)
y_pred

array([0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0,
       0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1,
       0, 0, 0])

In [10]:
print("\nModel Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))


Model Accuracy: 0.8324022346368715

Classification Report:
               precision    recall  f1-score   support

           0       0.89      0.87      0.88       126
           1       0.71      0.74      0.72        53

    accuracy                           0.83       179
   macro avg       0.80      0.80      0.80       179
weighted avg       0.83      0.83      0.83       179


Confusion Matrix:
 [[110  16]
 [ 14  39]]


In [11]:
g_model = GaussianNB()
g_model.fit(X_train, y_train)

y_pred = g_model.predict(X_test)
print(y_pred)

print("\nModel Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))

[0 1 1 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 1 0 0 0 1 1 0 1 0 1 0 0 0 0 0 0 0 1 0
 1 1 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 1
 1 1 1 0 0 0 0 1 0 0 0 1 0 1 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 1 0 0 0 0
 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 1 0 1 1 0 0 1 0 0 1 0 0 1 0 0 0
 0 1 1 1 1 1 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 1 1 0 1 0 0 0]

Model Accuracy: 0.8044692737430168

Classification Report:
               precision    recall  f1-score   support

           0       0.85      0.88      0.86       126
           1       0.69      0.62      0.65        53

    accuracy                           0.80       179
   macro avg       0.77      0.75      0.76       179
weighted avg       0.80      0.80      0.80       179


Confusion Matrix:
 [[111  15]
 [ 20  33]]


In [12]:
m_model = MultinomialNB(alpha=0.1)
m_model.fit(X_train, y_train)

y_pred = m_model.predict(X_test)
print(y_pred)

print("\nModel Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))

[0 1 1 0 0 0 0 0 0 1 0 0 0 0 0 1 0 1 1 0 0 1 1 1 0 0 0 1 0 0 0 0 0 0 0 1 0
 1 1 0 0 0 1 0 0 0 0 0 0 1 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1
 1 1 1 1 0 0 0 1 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1
 0 1 0 1 1 0 0 0 0 0 0 0 0 0 1 1 0 0 1 0 0 0 0 1 1 0 0 1 0 0 1 0 0 1 0 0 0
 0 1 1 0 1 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 1 0 1 0 0 0]

Model Accuracy: 0.8324022346368715

Classification Report:
               precision    recall  f1-score   support

           0       0.86      0.90      0.88       126
           1       0.74      0.66      0.70        53

    accuracy                           0.83       179
   macro avg       0.80      0.78      0.79       179
weighted avg       0.83      0.83      0.83       179


Confusion Matrix:
 [[114  12]
 [ 18  35]]


In [13]:
random_forest_model = RandomForestClassifier(random_state=42, n_estimators=50, max_depth=130, min_samples_split=2)
random_forest_model.fit(X_train, y_train)
y_pred = random_forest_model.predict(X_test)

print("\nModel Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))



Model Accuracy: 0.8212290502793296

Classification Report:
               precision    recall  f1-score   support

           0       0.91      0.83      0.87       126
           1       0.67      0.79      0.72        53

    accuracy                           0.82       179
   macro avg       0.79      0.81      0.80       179
weighted avg       0.83      0.82      0.83       179


Confusion Matrix:
 [[105  21]
 [ 11  42]]


In [14]:
decision_tree_model = DecisionTreeClassifier(random_state=42)
decision_tree_model.fit(X_train, y_train)

y_pred = decision_tree_model.predict(X_test)
print("\nModel Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))



Model Accuracy: 0.7150837988826816

Classification Report:
               precision    recall  f1-score   support

           0       0.82      0.76      0.79       126
           1       0.52      0.60      0.56        53

    accuracy                           0.72       179
   macro avg       0.67      0.68      0.67       179
weighted avg       0.73      0.72      0.72       179


Confusion Matrix:
 [[96 30]
 [21 32]]


In [15]:
param_grid = {
    "max_depth": [3, 5, 10, 15, None],
    "min_samples_split": [2, 5, 10, 15],
    "min_samples_leaf": [1, 2, 4, 8],
    "criterion": ["gini", "entropy"]
}

grid_search = GridSearchCV(DecisionTreeClassifier(random_state=42), param_grid, cv=10, scoring="accuracy")
grid_search.fit(X_train, y_train)

print("\nBest Parameters:", grid_search.best_params_)

best_model = DecisionTreeClassifier(random_state=42, **grid_search.best_params_)
best_model.fit(X_train, y_train)

y_pred = best_model.predict(X_test)

print("\nOptimized Model")
print("\nModel Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))


Best Parameters: {'criterion': 'gini', 'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 10}

Optimized Model

Model Accuracy: 0.7039106145251397

Classification Report:
               precision    recall  f1-score   support

           0       0.81      0.76      0.78       126
           1       0.50      0.57      0.53        53

    accuracy                           0.70       179
   macro avg       0.65      0.66      0.66       179
weighted avg       0.72      0.70      0.71       179


Confusion Matrix:
 [[96 30]
 [23 30]]


In [16]:
from joblib import dump

dump(m_model, "../models/multinomialNB_model_rs_42.sav")

['../models/multinomialNB_model_rs_42.sav']

Best Model

BernoulliNB and MultinomialNB have the same highest precision (0.8324). However, comparing the results for precision, recall and f1-score, MultinomialNB seems to be the ideal choice:

    MultinomialNB has better precision (0.86) and recall (0.90) for class 0 compared to BernoulliNB (precision 0.89 and recall 0.87).

    For class 1, MultinomialNB has better